README.md
# NoMorePass DART libraries

This library allows you to use the nomorepass.com system to send and
receive credentials using nomorepass mobile application. It is intended 
to use in any environment, so it does not generate / print the qr-code, 
instead provides the text that should be included in the qrcode 
(you can generate using any qrcode libraries).

## Usage

To receive passwords:

```
// Nomorepass initialization
  Nomorepass nmp = Nomorepass();
  nmp.apikey = 'MYAPIKEY';

  // QR to send a password
  print("Generating QR...");
  print(await nmp.getQrText('test'));
  print("Draw and scan this QR with nomorepass to send a password back");
  Map? res = null;
  Future.delayed(const Duration(seconds: 30), () {
    if (res == null) {
      nmp.stopped = true;
    }
  }).then((value) {
    if (nmp.stopped == true) {
      print("Aborted!");
      nmp.stop();
    }
  });
  print("Waiting for reception (30sec waiting)");
  nmp.stopped = false;
  res = await nmp.start();
  if (res != null && res.containsKey('error')) {
    print("Error ${res['error']}");
  } else {
    if (res != null) {
      print("User: ${res['user']}");
      print("Pass: ${res['password']}");
      print("Extra: ${res['extra']}");
    }
  }
```

To send password:

```
  // Nomorepass initialization
  Nomorepass nmp = Nomorepass();
  nmp.apikey = 'MYAPIKEY';

  print("Generating QR");
  print(await nmp.getQrSend("Test password", "usertest", "mypassword", {}));
  print("Scan this QR with nomorepass to receive a new password");
  Map? res = null;
  Future.delayed(const Duration(seconds: 30), () {
    if (res == null) {
      nmp.stopped = true;
    }
  }).then((value) {
    if (nmp.stopped == true) {
      print("Aborted!");
      nmp.stop();
    }
  });
  print("Waiting for reception (30sec waiting)");
  res = await nmp.send();
  print ("Received!);
```

## Examples

You have a exampe file (example.dart) in the example file

## How to use NoMorePass

1. Download and install the mobile app

* [android] https://play.google.com/store/apps/details?id=com.biblioeteca.apps.NoMorePass
* [ios] https://itunes.apple.com/us/app/no-more-pass/id1199780162?l=es&ls=1&mt=8

3. Open it and create a new password (or use some of yours)
4. Then you can scan the qrcode generated by the library to send securely this password to your app or send/update passwords from your code to the app.

## Help / more info

Visit [nomorepass.com](https://nomorepass.com) or leave an Issue

(C) 2021 Nomorepass.com + Biblioeteca Technologies


CHANGELOG.md
## 1.2.4

  - Includes generic KEY for nomorekeys
## 1.2.3

 - More Bug fixes
## 1.2.1

 - Bug fixes
## 1.2.0

 - Adds BLE Keys sending
## 1.1.0

 - Adds expiry to petitions (optional)
## 1.0.1

 - Includes extra fields on sendRemotePassToDevice (optional)
## 1.0.0

 - First version


example/ex2.dart
import '../lib/nomorepass.dart';

void main() async {
  Nomorepass nmp = Nomorepass();

  Map? res = null;
  int expiry =
      (new DateTime.now().microsecondsSinceEpoch / 1000000).round() + 200;
  nmp.setExpiry(expiry); // 200 sec expiry
  print('Expiry: $expiry');
  print("Generating QR...");
  String? url =
      await nmp.getQrNomorekeys('testKey', 'Daniel', '123456', 'KEY', {
    'host': 'puertaCasa',
    'extra': {
      'deviceid': 300,
      'external_id': 840,
      'shleft': 0,
      'type': 'padkey'
    }
  });
  //print(await nmp.getQrText('prueba'));
  print(url);
  print("Scan this QR to send a password back");
  // res = null;
  Future.delayed(const Duration(seconds: 60), () {
    if (res == null) {
      nmp.stopped = true;
    }
  }).then((value) {
    if (nmp.stopped == true) {
      print("Aborted!");
      nmp.stop();
    }
  });
  print("Waiting for reception (30sec waiting)");
  nmp.stopped = false;
  res = await nmp.send();
  if (res != null && res.containsKey('error')) {
    print("Error ${res['error']}");
  } else {
    if (res != null) {
      print("Password received:");
    }
  }
}


lib/src/crypto.dart
import 'dart:typed_data';
import 'package:encrypt/encrypt.dart';
import 'dart:convert';
import 'dart:math';
import 'package:crypto/crypto.dart';

class NmpCrypto {
  List<int> derive_key_and_iv(
      String password, List<int> salt, int key_length, int iv_lenght) {
    List<int> d = [];
    List<int> d_i = [];
    while (d.length < key_length + iv_lenght) {
      d_i = md5.convert(d_i + utf8.encode(password) + salt).bytes;
      d += d_i;
    }
    return d;
  }

  String encrypt(text, password) {
    var random = Random.secure();
    // Calculamos un random para salt (hasta 16 bytes)
    var salt = List<int>.generate(8, (i) => random.nextInt(256));
    final keyandiv = this.derive_key_and_iv(password, salt, 32, 16);

    final key = Key(Uint8List.fromList(keyandiv.sublist(0, 32)));
    final iv = IV(Uint8List.fromList(keyandiv.sublist(32, 32 + 16)));

    final encrypter = Encrypter(AES(key, mode: AESMode.cbc));

    List<int> res = utf8.encode('Salted__') + salt;
    final encrypted = encrypter.encrypt(text, iv: iv);
    res += encrypted.bytes;
    return base64.encode(res);
  }

  String decrypt(String text, String pass) {
    final input = base64.decode(text);
    final salt = input.sublist(8, 16);
    final keyandiv = this.derive_key_and_iv(pass, salt, 32, 16);
    final key = Key(Uint8List.fromList(keyandiv.sublist(0, 32)));
    final iv = IV(Uint8List.fromList(keyandiv.sublist(32, 32 + 16)));

    final encrypter = Encrypter(AES(key, mode: AESMode.cbc));
    final encrypted =
        Encrypted.fromBase64(base64Encode(input.sublist(16, input.length)));
    final decrypted = encrypter.decrypt(encrypted, iv: iv);

    return decrypted;
  }
}

lib/src/nomorepass.dart
import 'dart:math';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'crypto.dart';

class Nomorepass {
  String apikey = 'FREEAPI';
  String server = 'api.nomorepass.com';
  String base = 'https://api.nomorepass.com';
  String getidUrl = '';
  String checkUrl = '';
  String referenceUrl = '';
  String grantUrl = '';
  String pingUrl = '';
  bool stopped = false;
  String ticket = '';
  String token = '';
  NmpCrypto nmpc = new NmpCrypto();
  int? expiry = null;

  Nomorepass([String? server, String? apikey]) {
    server ??= 'api.nomorepass.com';
    apikey ??= 'FREEAPI';
    this.apikey = apikey;
    this.server = server;
    this.base = "https://" + server;
    this.getidUrl = this.base + "/api/getid.php";
    this.checkUrl = this.base + "/api/check.php";
    this.referenceUrl = this.base + "/api/reference.php";
    this.grantUrl = this.base + "/api/grant.php";
    this.pingUrl = this.base + "/api/ping.php";
  }

  String newToken([int? size]) {
    final _random = new Random();
    String charset =
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    if (size == null) size = 12;
    String ret = "";
    for (int i = 0; i < size; i++) {
      ret += charset[_random.nextInt(charset.length)];
    }
    return ret;
  }

  void setExpiry(int? expiry) {
    this.expiry = expiry;
  }

  Future<String?> getQrText(String site) async {
    Map<String, String> data = {'site': site};
    if (this.expiry != null) {
      data['expiry'] = this.expiry.toString();
    }
    var headers = {'User-Agent': 'NoMorePass-Dart/1.0', 'apikey': this.apikey};
    var url = Uri.parse(this.getidUrl);
    var resp = await http.post(url, headers: headers, body: data);
    if (resp.statusCode == 200) {
      final cuerpo = resp.body;
      final datos = json.decode(cuerpo);
      if (datos['resultado'] == 'ok') {
        this.ticket = datos['ticket'];
        this.token = this.newToken();
        final text = 'nomorepass://' + this.token + this.ticket + site;
        return text;
      }
    }
    return null;
  }

  Future<String?> getQrNomorekeys(String? site, String user, String password,
      String type, Map? extra) async {
    // Returns the QR url to send a nomorekeys key to the phone
    // basically the same as getQRSend but with nomorekeys://
    // only available for soundkey and lightkey right now
    // SOUNDKEY passwords are limited to 14 characters
    // LIGHTKEY are a unsigned int
    if (type != "SOUNDKEY" && type != "LIGHTKEY" && type != "BLEKEY") {
      type = "KEY";
    }
    if (site == null) {
      site = "WEBDEVICE";
    }
    Map<String, String> data = {'site': site};
    if (this.expiry != null) {
      data['expiry'] = this.expiry.toString();
    }
    var headers = {'User-Agent': 'NoMorePass-Dart/1.0', 'apikey': this.apikey};
    var url = Uri.parse(this.getidUrl);
    var resp = await http.post(url, headers: headers, body: data);
    if (resp.statusCode == 200) {
      final cuerpo = resp.body;
      final datos = json.decode(cuerpo);
      if (datos['resultado'] == 'ok') {
        final token = this.newToken();
        this.token = token;
        this.ticket = datos['ticket'];
        if (type == 'SOUNDKEY') {
          password = password.padRight(14).substring(0, 14);
        } else if (type == 'LIGHTKEY') {
          password = (int.parse(password) % 65536).toString();
        }
        final ep = this.nmpc.encrypt(password, token);
        String extrastr = '';
        if (extra != null) {
          if (!extra.containsKey('type')) {
            extra['type'] = type.toLowerCase();
          }
          if (extra.containsKey('extra')) {
            Map theextra = extra['extra']
                .map((key, value) => MapEntry(key, value?.toString()));
            if (theextra.containsKey('secret')) {
              extra['extra']['secret'] =
                  this.nmpc.encrypt(extra['extra']['secret'], token);
              if (!theextra.containsKey('type')) {
                extra['extra']['type'] = type.toLowerCase();
              }
            } else {
              if (!theextra.containsKey('type')) {
                theextra['type'] = type.toLowerCase();
              }
              extra['extra'] = theextra.map((key, value) => MapEntry(key,
                  (int.tryParse(value) == null) ? value : int.parse(value)));
            }
            extrastr = json.encode(extra);
          } else {
            extra = {
              'extra': {'type': type.toLowerCase()}
            };
            extrastr = json.encode(extra);
          }
        }
        final params = {
          'grant': 'grant',
          'ticket': this.ticket,
          'user': user,
          'password': ep,
          'extra': extrastr
        };
        print(params);
        url = Uri.parse(this.grantUrl);
        resp = await http.post(url, headers: headers, body: params);
        if (resp.statusCode == 200) {
          final dat = json.decode(resp.body);
          if (dat['resultado'] == 'ok') {
            final text = 'nomorekeys://' + type + token + dat['ticket'] + site;
            return text;
          }
        }
      }
    }
    return null;
  }

  Future<Map?> start() async {
    // Comenzamos a preguntar (check) si nos han enviado el pass
    // dado que esta función es síncrona no devolvemos el control
    // hasta que tenemos una respuesta (positiva o negativa)
    // o hasta que el valor del atributo stopped es cierto
    while (this.stopped == false) {
      final data = {'ticket': this.ticket};
      var headers = {
        'User-Agent': 'NoMorePass-Dart/1.0',
        'apikey': this.apikey
      };
      var url = Uri.parse(this.checkUrl);
      var resp = await http.post(url, headers: headers, body: data);
      if (resp.statusCode == 200) {
        final cuerpo = resp.body;
        //print(cuerpo);
        final decoded = json.decode(cuerpo);
        if (decoded['resultado'] == 'ok') {
          if (decoded['grant'] == 'deny') {
            return {'error': 'denied'};
          } else {
            if (decoded['grant'] == 'grant') {
              final res = {
                'user': decoded["usuario"],
                'password': nmpc.decrypt(decoded["password"], this.token),
                'extra': decoded['extra']
              };
              return res;
            } else {
              if (decoded['grant'] == 'expired') {
                return {'error': 'expired'};
              } else {
                await Future.delayed(Duration(seconds: 4));
              }
            }
          }
        } else {
          return {'error': decoded['error']};
        }
      } else {
        return {'error': 'network error'};
      }
    }
    this.stopped = false;
    return {'error': 'stopped'};
  }

  void stop() {
    this.stopped = true;
  }

  Future<String?> getQrSend(
      String? site, String user, String password, Map? extra) async {
    if (site == null) {
      site = "WEBDEVICE";
    }
    Map<String, String> data = {'site': site};
    if (this.expiry != null) {
      data['expiry'] = this.expiry.toString();
    }
    final headers = {
      'User-Agent': 'NoMorePass-Dart/1.0',
      'apikey': this.apikey
    };
    var url = Uri.parse(this.getidUrl);
    var resp = await http.post(url, headers: headers, body: data);
    if (resp.statusCode == 200) {
      final cuerpo = resp.body;
      final datos = json.decode(cuerpo);
      if (datos['resultado'] == 'ok') {
        final token = this.newToken();
        this.token = token;
        this.ticket = datos['ticket'];
        final ep = this.nmpc.encrypt(password, token);
        String extrastr = '';
        if (extra != null) {
          if (extra.containsKey('extra')) {
            Map theextra = extra['extra']
                .map((key, value) => MapEntry(key, value?.toString()));
            if (theextra.containsKey('secret')) {
              extra['extra']['secret'] =
                  this.nmpc.encrypt(extra['extra']['secret'], token);
              extra['extra']['type'] = 'pwd';
            } else {
              theextra['type'] = 'pwd';
              extra['extra'] = theextra.map((key, value) => MapEntry(key,
                  (int.tryParse(value) == null) ? value : int.parse(value)));
            }
            extrastr = json.encode(extra);
          } else {
            extra = {
              'extra': {'type': 'pwd'}
            };
            extrastr = json.encode(extra);
          }
        }
        final params = {
          'grant': 'grant',
          'ticket': this.ticket,
          'user': user,
          'password': ep,
          'extra': extrastr
        };
        url = Uri.parse(this.grantUrl);
        resp = await http.post(url, headers: headers, body: params);
        if (resp.statusCode == 200) {
          final dat = json.decode(resp.body);
          if (dat['resultado'] == 'ok') {
            final text = 'nomorepass://SENDPASS' + token + dat['ticket'] + site;
            return text;
          }
        }
      }
    }
    return null;
  }

  Future<Map?> send() async {
    // Comprueba si la contraseña enviada
    // se ha recibido. Solo para cuando
    // se recibe una respuesta positiva o negativa
    // o se comprueba que stopped = true
    while (this.stopped == false) {
      final params = {'device': 'WEBDEVICE', 'ticket': this.ticket};
      final headers = {
        'User-Agent': 'NoMorePass-Dart/1.0',
        'apikey': this.apikey
      };
      final url = Uri.parse(this.pingUrl);
      var resp = await http.post(url, headers: headers, body: params);
      if (resp.statusCode == 200) {
        final dat = json.decode(resp.body);
        if (dat['resultado'] == 'ok' && dat['ping'] == 'ok') {
          await Future.delayed(Duration(seconds: 4));
        } else {
          return dat;
        }
      }
    }
    return {"error": "stopped"};
  }

  Future<Map?> sendRemotePassToDevice(String? cloud, String deviceid,
      String secret, String username, String password, Map? extra) async {
    // Envía una contraseña remota a un dispositivo cloud
    // cloud: url de /extern/send_ticket
    // devideid: id del dispositivo
    // secret: md5 del secreto del dispositivo
    // username: usuario
    // password: contraseña
    String cloudurl = 'https://api.nmkeys.com/extern/send_ticket';
    if (cloud != null) {
      cloudurl = cloud;
    }
    final token = secret;
    var params = {'site': 'Send remote pass'};
    if (this.expiry != null) {
      params['expiry'] = this.expiry.toString();
    }
    final headers = {
      'User-Agent': 'NoMorePass-Dart/1.0',
      'apikey': this.apikey
    };
    var url = Uri.parse(this.getidUrl);
    var resp = await http.post(url, headers: headers, body: params);
    if (resp.statusCode == 200) {
      var dat = json.decode(resp.body);
      if (dat['resultado'] == 'ok') {
        final ticket = dat['ticket'];
        final ep = nmpc.encrypt(password, token);
        Map textra = {"type": "remote"};
        if (extra != null) {
          textra = extra;
        }
        params = {
          'grant': 'grant',
          'ticket': ticket,
          'user': username,
          'password': ep,
          'extra': json.encode(textra)
        };
        url = Uri.parse(this.grantUrl);
        resp = await http.post(url, headers: headers, body: params);
        if (resp.statusCode == 200) {
          dat = json.decode(resp.body);
          if (dat['resultado'] == 'ok') {
            params = {
              'hash': token.substring(0, 10),
              'ticket': ticket,
              'deviceid': deviceid
            };
            url = Uri.parse(cloudurl);
            resp = await http.post(url,
                headers: {'content-type': 'application/json'},
                body: utf8.encode(json.encode(params)));
            if (resp.statusCode == 200) {
              dat = json.decode(resp.body);
              return dat;
            } else {
              //print(resp.statusCode);
              return {"error": "error calling $cloudurl"};
            }
          } else {
            return {"error": dat};
          }
        } else {
          return {"error": "error calling granturl"};
        }
      } else {
        return {"error": dat};
      }
    } else {
      return {"error": "error calling getid"};
    }
  }
}


example/example.dart
import '../lib/nomorepass.dart';

void main() async {
  // Nomorepass inicialization
  Nomorepass nmp = Nomorepass();
  // Create token
  String texto = nmp.newToken(10);
  print("New token: $texto");

  // Crypto services
  NmpCrypto nmpc = NmpCrypto();
  final enc = nmpc.encrypt('lo que quiero encriptar', 'TOKEN');
  final dec = nmpc.decrypt(enc, 'TOKEN');
  print("Encrypted text: $enc");
  print("Decrypted text: $dec");

  // Remote pasword sending
  print("Sending password to remote device");
  print(await nmp.sendRemotePassToDevice(
      'https://test.nmkeys.com/extern/send_ticket',
      'DNCM91E3VO',
      '26bd80fd7c445933525c5768f58e1882',
      'test1remoto',
      'test1remoto', {
    'type': 'remote',
    'position': {'lat': 40.3488923, 'lon': -3.8079261}
  }));
  print("Sent!");
  // Password distribution
  // QR to send a key to mobile phone
  print("Generating QR");
  print(
      await nmp.getQrNomorekeys("TestSoundKey", "key", "Secret key", "BLEKEY", {
    'extra': {'secret': '1234567890123456'}
  }));
  print("Scan this QR to receive a new soundkey");
  print("Generating QR");
  print(await nmp.getQrSend("Test password", "usertest", "mypassword", {}));
  print("Scan this QR to receive a new password");
  Map? res = null;
  Future.delayed(const Duration(seconds: 30), () {
    if (res == null) {
      nmp.stopped = true;
    }
  }).then((value) {
    if (nmp.stopped == true) {
      print("Aborted!");
      nmp.stop();
    }
  });
  print("Waiting for reception (30sec waiting)");
  res = await nmp.send();
  // QR to send a password and expiry test
  int expiry =
      (new DateTime.now().microsecondsSinceEpoch / 1000000).round() + 20;
  nmp.setExpiry(expiry); // 20 sec expiry
  print("Generating QR...");
  print(await nmp.getQrText('prueba'));
  print("Scan this QR to send a password back");
  // res = null;
  Future.delayed(const Duration(seconds: 60), () {
    if (res == null) {
      nmp.stopped = true;
    }
  }).then((value) {
    if (nmp.stopped == true) {
      print("Aborted!");
      nmp.stop();
    }
  });
  print("Waiting for reception (30sec waiting)");
  nmp.stopped = false;
  res = await nmp.start();
  if (res != null && res.containsKey('error')) {
    print("Error ${res['error']}");
  } else {
    if (res != null) {
      print("User: ${res['user']}");
      print("Pass: ${res['password']}");
      print("Extra: ${res['extra']}");
    }
  }
}
